### PR TITLE
Implemented `knife user key list` and `knife client key list`.

### PR DESCRIPTION
**Pull request by tylercloke**
Abstracted out a bunch of common tests with the key create code to reduce redundancy.

Ping @chef/lob

### CODE REVIEW

**Comment by danielsdeleo** [comment]
@chef/client-core

File: spec/unit/knife/key_create_spec.rb [hunk 1]
```diff
@@ -75,26 +51,22 @@
...
-        allow(command).to receive(:service_object).and_return(service_object)
-        allow(command).to receive(:name_args).and_return(["charmander"])
-      end
-
-      context "when the command is successful" do
-        before do
-          expect(service_object).to receive(:run)
-        end
-      end
-    end
```
Comment (tylercloke) [inline spec/unit/knife/key_create_spec.rb]: All these deleted tests moved to common file to run against multiple classes.

File: lib/chef/knife/core/generic_presenter.rb [hunk 2]
```diff
@@ -216,7 +216,6 @@ def format_cookbook_list_for_display(item)
             end
           end
         end
-
```
Comment (jdmundrawala) [inline lib/chef/knife/core/generic_presenter.rb]: Can you undo this

File: lib/chef/knife/key_list.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+          max_length = 0
+          keys.each do |key|
+            key['name'] = key['name'] + ":"
+            max_length = key['name'].length if key['name'].length > max_length
+          end
+          keys.each do |key|
+            next if !key['expired'] && @config[:only_expired]
+            next if key['expired'] && @config[:only_non_expired]
+            display = "#{colorize(key['name']).ljust(max_length)} #{key['uri']}"
+            display = display + " (expired)" if key["expired"]
```
Comment (jdmundrawala) [inline lib/chef/knife/key_list.rb]: I like "#{display} (expired)" better

File: lib/chef/knife/key_list.rb [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+class Chef
+  class Knife
+    # Service class for UserKeyList and ClientKeyList,
+    # Implements common functionality of knife [user | org client] key list.
+    #
+    # @author Tyler Cloke
+    #
+    # @attr_accessor [Hash] cli input, see UserKeyList and ClientKeyList for what could populate it
+    class KeyList
```
Comment (jdmundrawala) [inline lib/chef/knife/key_list.rb]: nitpicky...you dont have to change this, but is the a list of keys or a thing that lists keys?

File: lib/chef/knife/core/generic_presenter.rb (hunk 2 above)
Comment (tylercloke) [inline lib/chef/knife/core/generic_presenter.rb]: Oops.

File: lib/chef/knife/key_list.rb (hunk 4 above)
Comment (tylercloke) [inline lib/chef/knife/key_list.rb]: A thing that lists keys. I can see how that would be confusing. I was following the convention I set with `KeyCreate` for the key creation service class.

**Comment by jdmundrawala** [comment]
:+1:

File: lib/chef/knife/key_list.rb (hunk 3 above)
Comment (btm) [inline lib/chef/knife/key_list.rb]: What about this? ;)
```
display << " (expired)" if key["expired"]
```

**Comment by btm** [comment]
:+1: 

I do wonder if the API allows listing keys without providing an actor, which could lead to something like ```knife user key list --all``` or even ```knife key list``` and ```knife key search```, but it looks like Chef::Key doesn't support that now anyway.

**Comment by tylercloke** [comment]
Yeah, the API is scoped to actors at the moment, but if that's something we would see as a useful feature in the future we should chat about it and look into prioritization :)

**Pull request merged**